game-core/src/test/java/games/strategy/triplea/delegate/battle/steps/retreat/OffensiveSubsRetreatTest.java

```diff
@@ -0,0 +1,195 @@
+package games.strategy.triplea.delegate.battle.steps.retreat;
+
+import static games.strategy.triplea.Constants.SUBMERSIBLE_SUBS;
+import static games.strategy.triplea.Constants.TRANSPORT_CASUALTIES_RESTRICTED;
+import static games.strategy.triplea.delegate.battle.FakeBattleState.givenBattleStateBuilder;
+import static games.strategy.triplea.delegate.battle.steps.BattleStepsTest.givenAnyUnit;
+import static games.strategy.triplea.delegate.battle.steps.BattleStepsTest.givenUnitCanEvade;
+import static games.strategy.triplea.delegate.battle.steps.BattleStepsTest.givenUnitDestroyer;
+import static games.strategy.triplea.delegate.battle.steps.BattleStepsTest.givenUnitTransport;
+import static org.mockito.ArgumentMatchers.any;
+import static org.mockito.ArgumentMatchers.anyBoolean;
+import static org.mockito.ArgumentMatchers.eq;
+import static org.mockito.Mockito.mock;
+import static org.mockito.Mockito.never;
+import static org.mockito.Mockito.verify;
+import static org.mockito.Mockito.when;
+
+import games.strategy.engine.data.GameData;
+import games.strategy.engine.data.Territory;
+import games.strategy.engine.data.properties.GameProperties;
+import games.strategy.engine.delegate.IDelegateBridge;
+import games.strategy.triplea.delegate.ExecutionStack;
+import games.strategy.triplea.delegate.battle.BattleActions;
+import games.strategy.triplea.delegate.battle.BattleState;
+import games.strategy.triplea.delegate.battle.MustFightBattle;
+import java.util.List;
+import org.junit.jupiter.api.extension.ExtendWith;
+import org.junit.jupiter.params.ParameterizedTest;
+import org.junit.jupiter.params.provider.Arguments;
+import org.junit.jupiter.params.provider.MethodSource;
+import org.mockito.Mock;
+import org.mockito.junit.jupiter.MockitoExtension;
+
+@ExtendWith(MockitoExtension.class)
+class OffensiveSubsRetreatTest {
+
+  @Mock ExecutionStack executionStack;
+  @Mock IDelegateBridge delegateBridge;
+  @Mock BattleActions battleActions;
+
+  @ParameterizedTest(name = "[{index}] {0}")
+  @MethodSource
+  void retreatHappens(final String displayName, final BattleState battleState) {
```

Comment:
nit, if you have to parameterize the display name, I think we might be parameterizing 'wrong'

The display name should likely be a constant for any test, the only thing that is changing are just values under test.

For example, if you are testing a 'sine' function, the input and expected outputs are parameterized, but in all cases we are testing a sine function. In part as well, JUnit will automatically print out the test parameters as part of the display.

To resolve, a comment on the test data could be sufficient. You could also consider passing in as parameters the components for a `Battlestate` and then constructing the battle state in the test.

Reply:
I tried your later idea (passing in parameters) but I couldn't build out a decent display name with the passed in parameters.

This does fit parameterizing, but the parameters are extremely complicated so building a human visible display name out of it is not workable.  I'll probably flip it so that the `retreatHappens` is just a helper `then` function and split out the parameterization into separate methods that call this `then` function.

Reply:
Codacity doesn't like what I did.  It is complaining that none of the test methods have `assert` or `fail`.  I can "fix" this by inlining the `then` methods or moving the `verify` calls from the `then` methods to the test methods.

Or, I could move back to the parameterized code and add a comment as you originally suggested.

@DanVanAtta , thoughts?

Reply:
I inlined the methods.  The `then` method was only three lines long.